Add parseSctpMap and writeSctpDescription functions

legastero opened the pull request.

fippo commented on sdp.js:
```diff
@@ -551,6 +551,39 @@ SDPUtils.parseMsid = function(mediaSection) {
+// receiving in this format
+SDPUtils.writeSctpDescription = function(media, sctp) {
+  return [
+    'm=' + media.kind + ' 9 ' + media.protocol + ' ' + sctp.protocol + '\r\n',
```
`m=application 9 DTLS/SCTP 5000`
-- chrome gives me this so sctp.protocol seems wrong and should be sctp.number?

legastero replied:
Oh, lovely. Firefox gives me `m=application 9 UDP/DTLS/SCTP webrtc-datachannel`

fippo replied:
urgh. When in doubt: @lgrahl, please advise

fippo replied:
i think the chrome format is an old one here...

legastero replied:
Yeah, that's an old version line. This PR will only emit the new format, but will ingest the old one.

lgrahl replied:
For maximum fun, `TCP/DTLS/SCTP` is also valid.

fippo replied:
github is clearly missing the :trollface: emoji reaction :-)

fippo commented on sdp.js:
```diff
@@ -551,6 +551,39 @@ SDPUtils.parseMsid = function(mediaSection) {
+// parses either the older sctpmap format, or the newer sctp-port format
+SDPUtils.parseSctpMap = function(mediaSection) {
+  var sctpMapLines = SDPUtils.matchPrefix(mediaSection, 'a=sctpmap:');
+  if (sctpMapLines.length > 0) {
```
try to parse the new format first?

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
+  if (sctpPort.length > 0) {
+    return {
+      number: sctpPort[0].substr(12),
+      protocol: 'webrtc-datachannel',
```
The protocol needs to be fetched from the `m`-line in the new format.

lgrahl commented on test/sdp.js:
```diff
@@ -832,6 +832,67 @@ describe('writeRtpDescription', () => {
+    });
+  });
+
+  const parsedNew = SDPUtils.parseSctpMap(newLine);
```
The output could change, so I'd suggest using a constant string instead.

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
+    return {
+      number: sctpPort[0].substr(12),
+      protocol: 'webrtc-datachannel',
+      streams: '1024'
```
We should not hand out this information at all. SCTP will negotiate and supersede this value in its handshake which is probably why it has been removed.

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
 
+// SCTP Datachannels
+// parses either the older sctpmap format, or the newer sctp-port format
+SDPUtils.parseSctpMap = function(mediaSection) {
```
```suggestion
SDPUtils.parseSctpDescription = function(mediaSection) {
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
   }
 };
 
+// SCTP Datachannels
```
```suggestion
// SCTP
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
+};
+
+// SCTP Datachannels
+// output only the sctp-port version now that all browsers support
```
```suggestion
// outputs the draft-ietf-mmusic-sctp-sdp-26 version that all browsers support by now
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,41 @@ SDPUtils.parseMsid = function(mediaSection) {
+  }
+};
+
+// SCTP Datachannels
```
```suggestion
// SCTP
```

legastero commented:
Tidied up suggestions, added max-message-size parsing.

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
 };
 
+// SCTP
+// parses either the older sctpmap format, or the newer sctp-port format
```
```suggestion
// parses draft-ietf-mmusic-sctp-sdp-26 first and falls back to draft-ietf-mmusic-sctp-sdp-05
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
+SDPUtils.parseSctpDescription = function(mediaSection) {
+  var mline = SDPUtils.parseMLine(mediaSection);
+  var maxSizeLine = SDPUtils.matchPrefix(mediaSection, 'a=max-message-size:');
+  var maxMessageSize = maxSizeLine.length > 0
```
I suggest:

```js
var maxMessageSize;
if (maxSizeLine.length > 0) {
  maxMessageSize = parseInt(maxSizeLine[0].substr(19), 10);
}
if (isNaN(maxMessageSize)) {
  maxMessageSize = 65536;
}
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
+    'c=IN IP4 0.0.0.0\r\n',
+    'a=sctp-port:' + sctp.number + '\r\n'
+  ];
+  if (sctp.maxMessageSize) {
```
```suggestion
  if (sctp.maxMessageSize !== undefined) {
```

`0` is valid and means that there is no restriction.

lgrahl commented on test/sdp.js:
```diff
@@ -832,6 +832,78 @@ describe('writeRtpDescription', () => {
+    parameters = {
+      protocol: 'webrtc-datachannel',
+      number: 5000,
+      maxMessageSize: '1024'
```
```suggestion
      maxMessageSize: 1024
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
+  var sctpPort = SDPUtils.matchPrefix(mediaSection, 'a=sctp-port:');
+  if (sctpPort.length > 0) {
+    return {
+      number: sctpPort[0].substr(12),
```
```suggestion
      port: sctpPort[0].substr(12),
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
+      .substr(10)
+      .split(' ');
+    return {
+      number: parts[0],
```
```suggestion
      port: parts[0],
```

lgrahl commented on sdp.js:
```diff
@@ -551,6 +551,50 @@ SDPUtils.parseMsid = function(mediaSection) {
+  var output = [
+    'm=' + media.kind + ' 9 ' + media.protocol + ' ' + sctp.protocol + '\r\n',
+    'c=IN IP4 0.0.0.0\r\n',
+    'a=sctp-port:' + sctp.number + '\r\n'
```
```suggestion
    'a=sctp-port:' + sctp.port + '\r\n'
```

lgrahl commented on test/sdp.js:
```diff
@@ -832,6 +832,78 @@ describe('writeRtpDescription', () => {
+  const parsedOld = SDPUtils.parseSctpDescription(oldSDP);
+  describe('parsing old form', () => {
+    it('parses port', () => {
+      expect(parsedOld.number).to.equal('5000');
```
```suggestion
      expect(parsedOld.port).to.equal('5000');
```

lgrahl commented on test/sdp.js:
```diff
@@ -832,6 +832,78 @@ describe('writeRtpDescription', () => {
+  const parsedNew = SDPUtils.parseSctpDescription(newSDP);
+  describe('parsing new form', () => {
+    it('parses port', () => {
+      expect(parsedNew.number).to.equal('5000');
```
```suggestion
      expect(parsedNew.port).to.equal('5000');
```

lgrahl commented on test/sdp.js:
```diff
@@ -832,6 +832,78 @@ describe('writeRtpDescription', () => {
+  beforeEach(() => {
+    parameters = {
+      protocol: 'webrtc-datachannel',
+      number: 5000,
```
```suggestion
      port: 5000,
```

legastero commented:
Applied suggestions. Also made the port parse as an int to match other port fields.

lgrahl commented:
Thanks. LGTM now. :+1:

legastero commented:
Squashed commits. All ready for you now @fippo

fippo commented on sdp.js:
```diff
@@ -551,6 +551,55 @@ SDPUtils.parseMsid = function(mediaSection) {
+};
+
+// SCTP
+// outputs the draft-ietf-mmusic-sctp-sdp-26 version that all browsers
```
I have another question on this one... when parsing the old format, could the following happen:
- browser receives old format from old server
- browser sends answer in new format

legastero replied:
It does look like Chrome and Firefox will still answer with the old format if that was used in the offer. So this change would silently remap it to the new format the old server can't handle. :/

Add an original format flag in the parsed sctp object?

fippo replied:
is there a property in the old format that isn't in the new format?

legastero replied:
There was the streams value that got removed during review. Otherwise the same. Unless you want to interpret media protocol of DTLS/SCTP as old format?

fippo replied:
that sounds like the best way to handle this

lgrahl replied:
Good catch @fippo. Yes, we could hand out and parse a `legacy` flag. A boolean might be risky though... you never know if they'll change it again... :wink:

legastero replied:
Does this mean I would need to parse and save the sctp streams value again to preserve for outputting this case? Or is it fine to drop in an answer?

lgrahl replied:
Nah, don't bother about parsing and just write `65535` (precisely this one, not `65536`) in the legacy format. I don't think anyone parses that since it's irrelevant after the SCTP handshake.

lgrahl commented:
Super-nit: Commit message needs updating.

legastero commented:
Ok, updated serializing to output sctpmap lines if the m-line protocol is DTLS/SCTP (without UDP/ or TLS/ prefix). Can go either way if that's good enough or if we want a legacy field.

Also fixed the prior commit message :)

lgrahl commented:
I'd personally prefer a boolean/enum *legacy* field, so it's independent on whether you have parsed a `media` line or not. Only if you guys agree though since it's easy enough to "hack".

fippo commented:
published as 2.10... now pondering what in adapter gets easier :-)
